### PR TITLE
Allow syncing custom notes

### PR DESCRIPTION
**Pull request by kanevk**
This PR introduces:
 - A change in Filename to Note title convention
 - Enables syncing new files on server run
 - Enables newly created remote notes to be synced as a file on the root of the notes root directory

**Affected issues:** #2

### CODE REVIEW

File: sync_note.py [hunk 1]
```diff
@@ -1,65 +1,53 @@
 import gkeepapi
-from pathlib import Path
+import os
```
Comment (scdekov) [inline sync_note.py:1]: The convention for imports structure is to have built in libs on top(preferably sorted),  after that separated with new line third party libs(I think only gkeepapi is third party here) and after that separated with new line imports from your project. And 2 new lines after imports

**Comment by kanevk** [comment]
#2

File: sync_note.py [hunk 2]
```diff
@@ -1,65 +1,53 @@
...
     newNote.labels.add(label)
-    #
+
     return newNote
 
 
 def find_note(keep, title):
-    label = keep.findLabel('autosync')
+    label = keep.findLabel(SYNC_LABEL)
     gnotes = list(
```
Comment (scdekov) [inline sync_note.py]: PO I would keep this on one line.

File: sync_note.py [hunk 3]
```diff
@@ -1,65 +1,53 @@
...
 
 
 def file_to_note_tuple(path):
-    path_after_home = path.replace(HOME, '')
-    title = path_after_home[1:].replace('/', ' ').replace('.md', '')
+    relative_path = path.replace(NOTES_ROOT, '')[1:]
+    title = relative_path.replace(f".{FILE_EXTENSION}", '')
+    print(f"title {title}")
+
     text = open(path, 'r').read()
```
Comment (scdekov) [inline sync_note.py]: you need to close the file descriptor here after finished reading. The easiest way is to use context manager
`
with open('path', 'r') as f:
    text = f.read()
`
this will call the __exit__ method of open after the block

File: sync_note.py [hunk 4]
```diff
@@ -73,7 +61,7 @@ def login(keep):
         token = keyring.get_password('google-keep-token',
```
Comment (scdekov) [inline sync_note.py:62]: No len and no == 0 is needed here.

Comment (scdekov) [inline sync_note.py:62]: Move getting of the token out of the if and else

File: sync_note.py [hunk 5]
```diff
@@ -106,14 +94,59 @@ def sync_down(keep):
         keep.sync()
```
Comment (scdekov) [inline sync_note.py:95]: Do you really need to check_connection? I think catching the exception below is good enough for connection problems and it'll also save you some time without this request to google.

File: sync_note.py [hunk 6]
```diff
@@ -106,14 +94,59 @@ def sync_down(keep):
...
+            # it'll create a file with the current version of the title which will be incomplete.
+            # During the next sync down a new file will be created for the same note with full title.
+            # This is a problem, because the *UPDATE* functionality its based on comparison of titles.
+            #
+            # In order to make it work for *CREATION*, we assume that when the user writes the title of a note,
+            # the content stays empty.
+            if not note.text:
+                continue
+
+            open(path, 'x').close()
```
Comment (scdekov) [inline sync_note.py:121]: what's the purpose of that?

File: sync_note.py [hunk 7]
```diff
@@ -106,14 +94,59 @@ def sync_down(keep):
...
+
+        # UPDATE
+        with open(path, 'r+') as file:
```
Comment (scdekov) [inline sync_note.py:124]: I think you don't need the "+" after "r"

File: sync_note.py [hunk 8]
```diff
@@ -106,14 +94,59 @@ def sync_down(keep):
...
+                file.write(note.text)
+
+
+@benchmark
+def upload_new_notes(keep):
+    all_paths = traverse_files(NOTES_ROOT)
+    file_note_states = list(
+        map(
+            file_to_note_tuple,
+            filter(lambda path: path[-3:] == f".{FILE_EXTENSION}", all_paths)
```
Comment (scdekov) [inline sync_note.py]: You can also use "string".endswith here

File: sync_server.py [hunk 9]
```diff
@@ -21,22 +23,22 @@ def on_modified(self, event):
...
-    observer.schedule(event_handler, path, recursive=True)
+    observer.schedule(event_handler, sync_note.NOTES_ROOT, recursive=True)
     observer.start()
     print("After observer start")
 
     timer = 0
     try:
         while True:
-            if timer % 60 == 0:
+            if timer % SYNC_DOWN_INTERVAL == 0:
```
Comment (scdekov) [inline sync_server.py]: Can't you just sleep for SYNC_DOWN_INTERVAL ?

File: utils.py [hunk 10]
```diff
@@ -9,5 +10,21 @@ def traverse_files(path):
...
+
+
+def benchmark(func):
+    """
+    A benchmark tool
+    """
+    def function_timer(*args, **kwargs):
+        start = time.time()
+        value = func(*args, **kwargs)
+        end = time.time()
```
Comment (scdekov) [inline utils.py]: PO shorten this to `print(f"{func.__name__} took {time.time() - start} seconds}")`

File: sync_note.py (hunk 6 above)
Comment (kanevk) [reply sync_note.py:121]: I need to create the file here.

Comment (scdekov) [reply sync_note.py:121]: But what is 'x' doing?

File: sync_note.py (hunk 5 above)
Comment (kanevk) [reply sync_note.py:95]: @scdekov We need it because the library `gkeepapi` makes the request in such a way that we receive some kind of Timeout error and another kind of error is raised in the same time.

File: sync_note.py (hunk 6 above)
Comment (kanevk) [reply sync_note.py:121]: `'x'` open for exclusive creation, failing if the file already exists

File: sync_note.py (hunk 7 above)
Comment (kanevk) [reply sync_note.py:124]: I need it since I'm both reading and writing.

Comment (scdekov) [reply sync_note.py:124]: Hmm, I thought that + is only creating the file if it doesn't already exist but that appeared to be completely wrong.

File: sync_note.py (hunk 4 above)
Comment (kanevk) [reply sync_note.py:62]: I'm getting rid of this in the next PR.

Comment (kanevk) [reply sync_note.py:62]: This will be deprecated, so I'm leaving it for now

File: Pipfile [hunk 11]
```diff
@@ -9,6 +9,7 @@ verify_ssl = true
 watchdog = "*"
 gkeepapi = "*"
 keyring = "*"
+pylint = "*"
```
Comment (scdekov) [inline Pipfile:12]: it's good practice to use a specific version here instead of *

File: Pipfile.lock [hunk 12]
```diff
@@ -1,7 +1,7 @@
 {
     "_meta": {
```
Comment (scdekov) [inline Pipfile.lock:2]: shouldn't this file be ignored?

File: sync_note.py [hunk 13]
```diff
@@ -1,65 +1,54 @@
...
-        value = func(*args, **kwargs)
-        end = time.time()
-        runtime = end - start
-        print(
-            f"`{func.__name__}` took {runtime} seconds"
-        )
-        return value
-    return function_timer
+
+CONFIG = json.load(open('.config.json'))
```
Comment (scdekov) [inline sync_note.py:13]: close the file after reading

File: sync_note.py [hunk 14]
```diff
@@ -1,65 +1,54 @@
...
-    path_after_home = path.replace(HOME, '')
-    title = path_after_home[1:].replace('/', ' ').replace('.md', '')
-    text = open(path, 'r').read()
-    #
+    relative_path = path.replace(NOTES_ROOT, '')[1:]
+    title = relative_path.replace(f".{FILE_EXTENSION}", '')
+
+    print(f"title {title}")
+
+    with open(path, 'r') as f: text = f.read()
```
Comment (scdekov) [inline sync_note.py:43]: PO don't inline like that

File: sync_note.py [hunk 15]
```diff
@@ -1,65 +1,54 @@
...
         raise Exception(f'Too many notes match the title: {title}')
     return (gnotes[0] if len(gnotes) == 1 else None)
 
 
 def file_to_note_tuple(path):
-    path_after_home = path.replace(HOME, '')
-    title = path_after_home[1:].replace('/', ' ').replace('.md', '')
-    text = open(path, 'r').read()
-    #
+    relative_path = path.replace(NOTES_ROOT, '')[1:]
```
Comment (scdekov) [inline sync_note.py:38]: that method returns dict, not tuple
